src/main/java/org/ice4j/util/PacketQueue.java

```diff
@@ -494,5 +501,332 @@ protected abstract T createPacket(
+            final long maxPackets = handler.maxPackets();
+
+            if (perNanos > 0 && maxPackets > 0)
+            {
+                final long now = System.nanoTime();
+                final long nanosRemainingTime = now - intervalStartTimeNanos;
+
+                if (nanosRemainingTime >= perNanos)
+                {
+                    intervalStartTimeNanos = now;
+                    packetsHandledWithinInterval = 0;
+                }
+                else if (packetsHandledWithinInterval >= maxPackets)
+                {
+                    return nanosRemainingTime;
+                }
+            }
+            return 0;
+        }
+
+        /**
+         * Count number of processed packets
+         */
+        void notifyPacketProcessed()
+        {
+            packetsHandledWithinInterval++;
+        }
+    }
+
+    /**
+     * Asynchronously reads packets from {@link #queue} on separate thread
+     * borrowed from {@link #executor}.
+     * Thread is not blocked when queue is empty and returned back to
+     * {@link #executor} pool. New or existing thread is borrowed from
+     * {@link #executor} when queue is non empty and {@link #reader} is not
+     * running
+     */
+    private final class AsyncPacketReader
+    {
+        /**
+         * Synchronization object of current instance state, in particular
+         * used to resolve races between {@link #schedule()} and {@link #reader}
+         * exit. In particular synchronization object used to access
+         * to field {@link #readerFuture} and {@link #delayedFuture}.
+         */
+        private final Object syncRoot = new Object();
+
+        /**
+         * Throttle calculator which is used for counting handled packets and
+         * computing necessary delay before processing next packet when
+         * throttling is enabled.
+         */
+        private final ThrottleCalculator<T> throttler
+            = new ThrottleCalculator<>(handler);
+
+        /**
+         * Stores <tt>Future</tt> of currently executing {@link #reader}
+         */
+        private Future<?> readerFuture;
+
+        /**
+         * Stores <tt>ScheduledFuture</tt> of currently delayed
+         * execution {@link #reader}
+         */
+        private ScheduledFuture<?> delayedFuture;
+
+        /**
+         * Perpetually reads packets from this {@link PacketQueue} and uses
+         * {@link #handler} on each of them.
+         */
+        private final Runnable reader = new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                final long maxSequentiallyProcessedPackets =
+                    handler.maxSequentiallyProcessedPackets();
+
+                int sequentiallyHandledPackets = 0;
+
+                while (!closed.get())
+                {
+                    T pkt;
+
+                    synchronized (syncRoot)
+                    {
+                        long delay = throttler.getDelayNanos();
```

Comment:
i'm only taking some quick looks here, but i don't quite get the throttler.  i thought the idea was:

read from the queue and if:
1) it was empty -> this context does nothing until it is rescheduled via something being added to the queue
2) there is data, and we're within max sequential packets -> process the data
3) there is data, and we're now beyond max sequential packets -> wait some delay before trying to read again

what case is the throttler handling that max sequential isn't?

Reply:
@bbaldino the notion of throttler was borrowed [from `libjitsi`](https://github.com/jitsi/libjitsi/blob/6c0a31d6193dcd2b012ae3ced656ad403403d9ff/src/org/jitsi/impl/neomedia/RTPConnectorOutputStream.java#L937). I've added it here because I'm going to replace class in `libjitsi` by this `PacketQueue` from `ice4j`.
`maxSequentiallyProcessedPackets` is for implementing cooperative multi-tasking: for example, when some queue  always has something to proceed and it shares executor thread with some other queue then it is important to switch execution between different queues, so all queues are processed and there is no queues which suffer from execution starvation. 
Throttling and `maxSequentiallyProcessedPackets` could be turned on/off independently of each other, so it's quite fine to do both checks in this while loop to achieve both throttling and cooperative multi tasking.

Reply:
My understand of throttling is following: queue should not proceed more than `N` items per `P` seconds. For example, it could be configured that there should be no more than 100 processed items per second, which is completely different from max items processed in row, which is controlled by `maxSequentiallyProcessedPackets `.

Reply:
Ok, I understand.  I'm not sure if we really need that, and would be great to simplify this (I understand it can be disabled, but it does add complexity).  For what it's worth, the usage of this will be gone in the bridge rewrite, but I would love to be able to use this scheme on a LinkedBlockingQueue, as we discussed.  I'm not sure we necessarily need to block this work on the rewrite, but, maybe we can at least make AsyncPacketReader a first class citizen (a class in its own file) as I think it's a useful class.